File: maps/src/main/java/it/localhostsoftware/maps/model/Marker.java
package it.localhostsoftware.maps.model;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public abstract class Marker<M> {
    private final M m;

    public Marker(M m) {
        this.m = m;
    }

    public M getMarker() {
        return m;
    }

    abstract public void remove();

    abstract public String getId();

    abstract public void setPosition(@NonNull LatLng<?> var1);

    abstract public LatLng<?> getPosition();

    abstract public void setZIndex(float var1);

    abstract public float getZIndex();

    abstract public void setIcon(@Nullable BitmapDescriptor<?> var1);

    abstract public void setAnchor(float var1, float var2);

    abstract public void setInfoWindowAnchor(float var1, float var2);

    abstract public void setTitle(@Nullable String var1);

    abstract public String getTitle();

    abstract public void setSnippet(@Nullable String var1);

    abstract public String getSnippet();

    abstract public void setDraggable(boolean var1);

    abstract public boolean isDraggable();

    abstract public void showInfoWindow();

    abstract public void hideInfoWindow();

    abstract public boolean isInfoWindowShown();

    abstract public void setVisible(boolean var1);

    abstract public boolean isVisible();

    abstract public void setFlat(boolean var1);

    abstract public boolean isFlat();

    abstract public void setRotation(float var1);

    abstract public float getRotation();

    abstract public void setAlpha(float var1);

    abstract public float getAlpha();

    abstract public void setTag(@Nullable Object var1);

    @Nullable
    abstract public Object getTag();
}

File: maps/src/main/java/it/localhostsoftware/maps/model/Marker.kt
package it.localhostsoftware.maps.model

import android.content.Context
import com.google.android.gms.maps.model.MarkerOptions
import it.localhostsoftware.maps.MobileServices
import it.localhostsoftware.maps.getMobileServices
import it.localhostsoftware.maps.google.model.GoogleMarkerOptions
import it.localhostsoftware.maps.huawei.model.HuaweiMarkerOptions

abstract class Marker<M, LL : LatLng<*>, BD : BitmapDescriptor<*>>(val m: M) {
    abstract fun remove()
    abstract val id: String
    abstract var position: LL
    abstract var zIndex: Float
    abstract fun setIcon(var1: BD?)
    abstract fun setAnchor(var1: Float, var2: Float)
    abstract fun setInfoWindowAnchor(var1: Float, var2: Float)
    abstract var title: String?
    abstract var snippet: String?
    abstract var isDraggable: Boolean
    abstract fun showInfoWindow()
    abstract fun hideInfoWindow()
    abstract val isInfoWindowShown: Boolean
    abstract var isVisible: Boolean
    abstract var isFlat: Boolean
    abstract var rotation: Float
    abstract var alpha: Float
    abstract var tag: Any?
}

abstract class MarkerOptions<MO, BD : BitmapDescriptor<*>, LL : LatLng<*>>(val mo: MO) {
    companion object {
        fun getInstance(c: Context) =
            when (c.getMobileServices()) {
                MobileServices.GOOGLE -> GoogleMarkerOptions(MarkerOptions())
                MobileServices.HUAWEI -> HuaweiMarkerOptions(com.huawei.hms.maps.model.MarkerOptions())
                else -> throw IllegalStateException()
            }
    }

    abstract var position: LL
    abstract var title: String?
    abstract var snippet: String?
    abstract var icon: BD?
    abstract var anchor: Pair<Float, Float>
    abstract var isDraggable: Boolean
    abstract var isVisible: Boolean
    abstract var isFlat: Boolean
    abstract var rotation: Float
    abstract var infoWindowAnchor: Pair<Float, Float>
    abstract var alpha: Float
    abstract var zIndex: Float
}